merge/merge.go

```diff
@@ -0,0 +1,171 @@
+package diff
+
+import (
+	"log"
+	"strconv"
+)
+
+const indicesReorderedKey = "$"
+
+// Merge merges an update diff into the previous version of a JSON object,
+// and returns the updated version.
+//
+// It effectively reverses the diff algorithm implemented in package diff.
+func Merge(prev map[string]interface{}, diff map[string]interface{}) interface{} {
+	new := make(map[string]interface{})
+
+	// Update exisiting fields.
+	for k, v := range prev {
+		d, ok := diff[k]
+		// No change.
+		if !ok {
+			new[k] = v
+			continue
+		}
+		// Removed.
+		if isRemoved(d) {
+			continue
+		}
+		// Updated.
+		var newV interface{}
+		switch v := v.(type) {
+		case map[string]interface{}:
+			newV = MergeMap(v, d)
+		case []interface{}:
+			newV = MergeArray(v, d)
+		default:
+			newV = MergeReplaced(d)
```

Comment:
There's a lot of calls to MergeReplaced. I wonder if we could simplify the overall logic and replace those individual checks with single check before the switch statement that is "is this a slice, if so, this is a replacement".